server/test/integration/com/thoughtworks/go/server/security/ReAuthenticationFilterTest.java

```diff
@@ -0,0 +1,137 @@
+    private HttpSession session;
+    private HttpServletRequest request;
+    private HttpServletResponse response;
+    private FilterChain filterChain;
+    private TimeProvider timeProvider;
+
+    @Before
+    public void setUp() throws Exception {
+        timeProvider = mock(TimeProvider.class);
+        session = mock(HttpSession.class);
+        request = mock(HttpServletRequest.class);
+        response = mock(HttpServletResponse.class);
+        filterChain = mock(FilterChain.class);
+        systemEnvironment = mock(SystemEnvironment.class);
+        filter = new ReAuthenticationFilter(systemEnvironment, timeProvider);
+
+        stub(request.getSession()).toReturn(session);
+        stub(request.getRequestedSessionId()).toReturn("session_id");
+    }
+
+    @Test
+    public void shouldContinueWithChainAndReturnIfReAuthenticationIsDisabled() throws IOException, ServletException {
+        when(systemEnvironment.isReAuthenticationEnabled()).thenReturn(false);
+
+        filter.doFilterHttp(request, response, filterChain);
+
+        verify(filterChain).doFilter(request, response);
+        verifyNoMoreInteractions(filterChain);
+    }
+
+    @Test
+    public void shouldContinueWithChainAndReturnIfNotAuthenticated() throws IOException, ServletException {
+        when(systemEnvironment.isReAuthenticationEnabled()).thenReturn(true);
+
+        filter.doFilterHttp(request, response, filterChain);
+
+        verify(filterChain).doFilter(request, response);
+        verifyNoMoreInteractions(filterChain);
+    }
+
+    @Test
+    public void shouldContinueWithChainAndReturnForAuthenticatedSessionWithoutLastAuthenticationTimeStamp() throws IOException, ServletException {
+        long currentTimeMillis = DateTimeUtils.currentTimeMillis();
+        Authentication authentication = setupAuthentication();
+
+        when(systemEnvironment.isReAuthenticationEnabled()).thenReturn(true);
+        when(timeProvider.currentTimeMillis()).thenReturn(currentTimeMillis);
+
+        filter.doFilterHttp(request, response, filterChain);
+
+        verify(session).setAttribute("last_reauthentication_check_time", currentTimeMillis);
+        verify(filterChain).doFilter(request, response);
+        verifyNoMoreInteractions(filterChain);
+        assertTrue(authentication.isAuthenticated());
+    }
+
+    @Test
+    public void shouldReAuthenticateIfReAuthTimeIntervalHasElapsed() throws IOException, ServletException {
+        long currentTimeMillis = DateTimeUtils.currentTimeMillis();
+        long minuteBack = DateTimeUtils.currentTimeMillis() - 60000;
+        Authentication authentication = setupAuthentication();
+
+        when(timeProvider.currentTimeMillis()).thenReturn(currentTimeMillis);
+        when(systemEnvironment.isReAuthenticationEnabled()).thenReturn(true);
+        when(systemEnvironment.reAuthenticationTimeInterval()).thenReturn(55000L);
+        when(session.getAttribute("last_reauthentication_check_time")).thenReturn(minuteBack);
```

Comment:
can we extract this into constant  `LAST_REAUTHENICATION_CHECK_TIME` or use the same constant as of the filter?